### PR TITLE
Archive storage config API

### LINKED ISSUE #336: Add default storage config to bucket archive api
This would be the default that applies to all bucket archives per account/user.

### LINKED ISSUE #339: Add per bucket archive storage config option
Add ability to provide a custom storage config for a single call to `Buckets.Archive`

### PR DESCRIPTION
**Pull request by asutula**
Allows getting and setting the default `ArchiveConfig` per `Bucket` and allows you to pass an override `ArchiveConfig` with each call to `Archive`.

Updates everything from mongo, rpc, client, local buckets api, and cli.

closes #336 
closes #339

### CODE REVIEW

File: api/buckets/client/client.go
Comment (asutula) [inline api/buckets/client/client.go:366]: Added options arg to specify a custom `ArchiveConfig`. This is also reflected in the local Buckets API.

File: api/buckets/pb/buckets.proto
Comment (asutula) [inline api/buckets/pb/buckets.proto:218]: Added the custom override config here. This can be `nil` when the default should be used.

File: api/buckets/pb/buckets.proto
Comment (asutula) [inline api/buckets/pb/buckets.proto:286]: New methods for getting and setting the bucket default archive config.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:74]: This is the base `StorageConfig` we use to store bucket data in Powergate. Biggest thing to note here is that hot storage is disabled. When a bucket is archived, we create a `ffs.FilConfig` from a `ArchiveConfig` and set `Cold.Filecoin` value with it.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:83]: If the user hasn't set a default `ArchiveConfig` for their Bucket, this one is used. I'd like to export the default `StorageConfig` from Powergate so we can create this dynamically based on whatever is set in Powergate.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2042]: Just use the default specified at the top of this file. Logic of not persisting a default on the `BucketArchive` struct is that maybe we want to change the default with time. Don't want to be stuck with an old default that was persisted to mongo. We should take this approach in Powergate @jsign .

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2039]: Use the `ArchiveConfig` that was set by the user if it exists.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2062]: If the user calls `SetDefaultArchiveConfig` explicitly, we do persist the `ArchiveConfig` they provide and it will be used as the default in `DefaultArchiveConfig` and `Archive` from now on.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2157]: Use the override if it's provided in the request.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2159]: Or use the user default.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2163]: Or just use the default default that is specified in code at the top of this file.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2167]: Build the `StorageConfig` by using the base value plus a `FilConfig` created from the `ArchiveConfig`

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2172]: The default `ArchiveConfig` doesn't include any address information, and if the user sets their own Bucket default `ArchiveConfig`, it may not have an address. In that case, grab the default address from the FFS instance. This is nice because it can change if the user wants. 

Long term, I'm working on a Powergate branch where default address in managed separately from default config, and address can be provided as an option in `PushConfig`, `Replace`, etc. That will make this cleaner.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2225]: Logic here, in the case of a second (or more) archive, is to get the current `StorageConfig` and compare it to the new `StorageConfig`. If they are the same, use `Replace`. If not, use `Remove` and `PushConfig`.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2239]: Here's that comparison. Cool to use `reflect`?

File: buckets/local/archive.go
Comment (asutula) [inline buckets/local/archive.go:140]: Adds option for override `ArchiveConfig`

File: buckets/local/archive.go
Comment (asutula) [inline buckets/local/archive.go:120]: This wasn't used.

File: go.mod
Comment (asutula) [inline go.mod:58]: Note: updated to newish version I think is being deployed everywhere

File: mongodb/bucketarchives.go
Comment (asutula) [inline mongodb/bucketarchives.go:13]: This will be `nil` until the user sets it. We take advantage of that fact to detect `nil` and use the hard coded default instead.

**Comment by asutula** [comment]
Ok getting a legit error in the integration test `TestArchiveBucketWorkflow`. The error is:
```
executing cold-storage config: making deal configs: getting miners from minerselector: not enough miners that satisfy the constraints
```
@jsign just checking if there is something we need to change about the lotus devnet and/or powergate setup in here? Anything come to mind?

**Comment by asutula** [comment]
Updated the compose file to powergate 0.6.6 and lotus-devnet sha-35afaa3 and now the error is:
```
executing cold-storage config: all proposals were rejected
```

**Comment by jsign** [comment]
> Ok getting a legit error in the integration test `TestArchiveBucketWorkflow`. The error is:
> 
> ```
> executing cold-storage config: making deal configs: getting miners from minerselector: not enough miners that satisfy the constraints
> ```
> 
> @jsign just checking if there is something we need to change about the lotus devnet and/or powergate setup in here? Anything come to mind?

Yes, most probably is related to forcing Peter's miners in the version of Powergate you've updated. Now that in v0.6.6 we switched to SR2 miner selector by default, might make sense again to have the original _StorageConfig_ which doesn't have a replication factor of 5, which is my best guess about that error. 
So the idea would be:
- Revert to the original _StorageConfig_ with rep facto 1 and without trusted miners.
- Set explicitly for tests here to use `--ffsminerselector=reputation`, since now the default is `sr2`.

I can confirm this theory on Monday, but makes sense to me.

**Comment by asutula** [comment]
Thanks @jsign. That gives me a path forward. Will dig in more later.

File: api/buckets/service.go
Comment (asutula) [inline api/buckets/service.go:2246]: @jsign I just want to check that this is ok. I have to disable hot and cold storage before calling `Remove`. I do that here, just ignore the jobId, and immediately call `Remove`. This works when I run the tests, but just want to make sure it's not just luck. Do I need to watch the job and wait until it is successful before calling `Remove`?

**Review by jsign — APPROVED** [review]
LGTM! I left some questions in the wild.

File: api/buckets/service.go
Comment (asutula) [reply api/buckets/service.go:74]: Correct. The `ArchiveConfig` that they can edit maps to `StorageConfig.Cold.FilConfig` so that is the only part of the `StorageConfig` they can effectively edit. They can't edit the `Hot` settings at all. `Hot` settings are exclusively controlled right here.

File: api/buckets/service.go
Comment (asutula) [reply api/buckets/service.go:77]: Yea I understand that. I decided to just build the long term solution here and figured we could discuss if/what to change in the short term. Soooo....

Should we remove `RepFactor`, `ExcludedMiners`, `TrustedMiners`, and `CountryCodes` from `ArchiveConfig` for now?

File: api/buckets/service.go
Comment (asutula) [reply api/buckets/service.go:2041]: No very strong option here either. It made sense in my mind for `ArchiveConfig` just because it is a new concept and it feels good to not commit too much to some default that is persisted and force us into having to do a data migration soon when things change. We can discuss further for Powergate if it ever seems important.

File: api/buckets/service.go
Comment (asutula) [reply api/buckets/service.go:2238]: I had read that `cmp.Equal` is meant for testing and panics under some circumstances. Not sure if that is true, didn't dig much deeper.

File: cmd/buck/cli/archive.go
Comment (sanderpick) [inline cmd/buck/cli/archive.go:16]: Just noticing now that these archive commands (the older ones as well) as missing the `Args` param, which is handy to restrict args, e.g., `Args: cobra.ExactArgs(0),`, `Args: cobra.RangeArgs(1, 2),`, `Args:  cobra.MaximumNArgs(1),`, etc. It also shows in the `--help` output.

**Review by sanderpick — APPROVED** [review]
LGTM! I left a bunch of nits, but otherwise looks solid.

File: api/buckets/service.go
Comment (asutula) [reply api/buckets/service.go:2245]: I think that for the end state that you want to exist the answer is: yes, we are on the same page about that. Hot storage will be disabled.

Just to clear some details up, in the scenario you describe, `oldStorageConfig` won't be `nil`, it will be whatever the previous bucket archive used, probably has hot storage enabled, etc. The new `storageConfig` will likely be different because it must have hot storage disabled, and likely has other differences because of the new default settings. So for that reason, we will `Remove` and `Push` the new storage config with hot storage disabled.

File: api/buckets/service.go
Comment (asutula) [reply api/buckets/service.go:2246]: It is disabled once this PR is merged in. But for previously archived buckets that are archiving again once this PR is merged (the situation we discuss above), hot storage may be enabled and will be disabled by this `PushStorageConfig`

File: integrationtest/pg/pg_test.go
Comment (asutula) [reply integrationtest/pg/pg_test.go:105]: Oops, did that to speed up local test debugging, will re-enable.

File: api/buckets/service.go
Comment (jsign) [reply api/buckets/service.go:78]: I'm not sure, so I'd leave that up to you. :)

File: api/buckets/service.go
Comment (jsign) [reply api/buckets/service.go:2239]: `cmp.Equal` was created as a better solution than `reflect.DeepEqual` which compares unexported fields, and might not behave correctly on particular types which its equality might not be easily comparable by bytes

I'd say both are meant for a quick way of comparing things. The full way is manual.

File: api/buckets/service.go
Comment (jsign) [reply api/buckets/service.go:2246]: 👍🏼

File: integrationtest/pg/docker-compose.yml
Comment (asutula) [reply integrationtest/pg/docker-compose.yml:38]: I think it was missing a space before so this is a formatting fix

File: api/buckets/service.go
Comment (asutula) [reply api/buckets/service.go]: Cool, switched to `cmd.Equal`

**Pull request merged**
